Added Solution Design Document

codepossible opened the pull request:
## Primary Changes
 - Updated Solution Design document
 - Added Message Schema document
 - Added JSON schemas and Sample for entities in CosmosDB and events.

codepossible committed:
Added Solution Design and Message Schema

codepossible committed:
Markdown lint

codepossible committed:
Updated reference link

davidkpiano commented on docs/message-schema-definitions.md:
```diff
@@ -0,0 +1,166 @@
+
+Link: [Shipment Example](./json/samples/input-shipment-sample.json)
+
+### Master Data Container (MDC)
```
```suggestion
### Master Data Collection (MDC)
```

davidkpiano replied:
```suggestion
### Master Data Collection (MDC)
```

davidkpiano commented on docs/message-schema-definitions.md:
```diff
@@ -0,0 +1,166 @@
+
+## Entity Schema
+
+Each entity represents the current state for a given store. It includes a dictionary of all products in the store, each of which will be updated by incoming events and subsequently written to the Cosmos MDS collection.
```
```suggestion
Each entity represents the current state for a given store. It includes a dictionary of all products in the store (keyed by product `id`), each of which will be updated by incoming events and subsequently written to the Cosmos MDS collection.
```

davidkpiano commented on docs/solution-design.md:
```diff
@@ -6,6 +6,50 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+- Azure Functions
+  - Change Feed Listeners
+  - Durable Function Orchestrator
+  - Durable Entity as a Virtual Actor
```
```suggestion
  - Durable Entity (similar to a Virtual Actor)
```

jcocchi replied:
Might want to clarify the durable entity is at the store level (i.e. there will be numStores durable entities)

codepossible replied:
Ok, will do that.

davidkpiano commented on docs/solution-design.md:
```diff
@@ -6,6 +6,50 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+
+## Next Step
+
+Now that the design is of the solution is clear, let's jump on to [Challenge 0](challenge-000.md), set up your development environment.
```
```suggestion
Now that the design of the solution is clear, let's jump on to [Challenge 0](challenge-000.md), set up your development environment.
```

jcocchi commented on docs/json/samples/event-payload-onHand-sample.json:
```diff
@@ -0,0 +1,12 @@
+    "storeId": "store005",
+    "upc": "0003400029005",
+    "countAdjustment": 1000,
+    "type": "onhandUpdate",
```
We just changed the event type to be camel case uniform with the rest of the project so it should now read
`"type": "onHandUpdate"`

codepossible replied:
Ok I will update that.

codepossible replied:
Done

jcocchi commented on docs/json/schemas/event-schema.json:
```diff
@@ -0,0 +1,75 @@
+        "type": {
+            "$id": "#/properties/type",
+            "type": "string",
+            "enum": ["shipmentUpdate", "onhandUpdate"],
```
We just changed the event type to be camel case uniform with the rest of the project so it should now read
`"enum": ["shipmentUpdate", "onHandUpdate"],`

codepossible replied:
Ok, I will update that.

codepossible replied:
Done

jcocchi commented on docs/message-schema-definitions.md:
```diff
@@ -0,0 +1,166 @@
+    "storeId": "store005",
+    "upc": "0003400029005",
+    "countAdjustment": 1000,
+    "type": "onhandUpdate",
```
changed to "onHandUpdate"

codepossible replied:
Updated

jcocchi commented on docs/solution-design.md:
```diff
@@ -6,6 +6,50 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+The following components will be used as part of the solution:
+  
+- Cosmos DB Account
+- Cosmos DB Containers
```
These containers are in a "inventory" database

```
Cosmos DB Account
    Cosmos DB Database (inventory)
        Cosmos DB Containers
           - On Hand (onHand)
           - Shipments (shipments)
           - Master Data Container (mdc)
```

codepossible replied:
Got it. I will make the change.

codepossible replied:
Done

jcocchi commented on docs/solution-design.md:
```diff
@@ -6,6 +6,50 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+  - Shipments (shipments)
+  - Master Data Container (mdc)
+- Azure Functions
+  - Change Feed Listeners
```
Might be worth it to enumerate what these will be (On Hand and Shipments) to clarify there is no change feed listener for Master Data

codepossible replied:
Added text in Azure Functions to make it clear that change feed listeners are only on the On Hand and Shipment CosmosDB containers.

jcocchi commented on docs/solution-design.md:
```diff
@@ -6,6 +6,50 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+
+The processing starts when documents are fed in to the Cosmos DB containers (shipments and on hand), which triggers a change feed in CosmosDB.
+
+Azure Functions configured to listen to the feed get a list of documents which are changed (add/edited/deleted). The Change feed listener convert them to events.
```
Change feed doesn't respond to document deletes, if we want to add this we will have to add a "soft delete" property and use the ttl to expire the documents

codepossible replied:
I will limit to the case we are using it for.

jcocchi commented on docs/solution-design.md:
```diff
@@ -6,6 +6,50 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+
+The events are sent to Durable Entity orchestrator, which routes these event based on store identity to the virtual actor representing the store.
+
+The virtual actor processes the event and updates its states. It then commits the data to the Master Data Store container in Cosmos DB.
```
You have been calling this "Master Data Container" in other places, should pick one name to use everywhere

codepossible replied:
We need to make it consistent. Master Data Collection (MDC) is the business terminology. In the new CosmosDB terminology, we create "Containers" instead of "Partitioned Collections". So I was trying to keep up with that.

jcocchi replied:
"Container" is the generic term for any Cosmos DB API and "Collection" is specific to SQL/ Mongo as "Graph" is specific to Gremlin and "Table" is specific to Cassandra. In this case either Container or Collection would be correct, just needs to be consistent

codepossible replied:
Thanks for the clarification @jcocchi

codepossible committed:
Update docs/message-schema-definitions.md

Co-authored-by: David Khourshid <davidkpiano@gmail.com>

codepossible committed:
Update docs/message-schema-definitions.md

Co-authored-by: David Khourshid <davidkpiano@gmail.com>

codepossible committed:
Update docs/solution-design.md

Co-authored-by: David Khourshid <davidkpiano@gmail.com>

codepossible committed:
Update docs/solution-design.md

Co-authored-by: David Khourshid <davidkpiano@gmail.com>

codepossible committed:
Updated to resolve PR comments

codepossible committed:
Merge branch 'shbanerj/42' of https://github.com/cse-southeast-devcrew/openhack-durable-entities into shbanerj/42

codepossible commented:
Updated based on Review Comments provided by @jcocchi  and @davidkpiano

codepossible committed:
Updated images with new naming convention

codepossible committed:
added non-dictionary words to spell check

brandonh-msft commented on docs/message-schema-definitions.md:
```diff
@@ -0,0 +1,166 @@
+
+### Shipments
+
+This collection stores one document for each shipment delivered to any store. It is partitioned by the `id` property, which is a GUID representing the shipment id. A change feed listens to this collection and will get documents that have been changed or updated in batches. The schema of these incoming documents matches the schema of the original defined below. The change feed processor will send  ...
```
so for N documents we have N partitions? should we have a better partition key? typically don't we desire for a partition to span a (somewhat equal-sized) group of items?

jcocchi replied:
Hmm looks like my comment didn't post.. sorry about that! 

For Cosmos DB `id` is usually a really good choice for partition key because it has a very high cardinality (1:1) and will evenly spread your data across all partitions. Since the workload for this collection is write heavy and doesn't really get queried, we just need to focus on spreading out our writes.

brandonh-msft commented on docs/message-schema-definitions.md:
```diff
@@ -0,0 +1,166 @@
+
+### Master Data Collection (MDC)
+
+This collection stores one document for each product in every store. It is partitioned by the `id` property, which is composed of `divId:storeId:upc`. This document is created or updated by the Orchestrator and its corresponding store entity.
```
consider partitioning by `divId` or `divId:storeId` instead?

brandonh-msft replied:
since the actual property name is `divisionId` - spell it out fully here so as not to confuse

jcocchi replied:
Comment didn't post here either...

`divId` only has 10 unique values, `divId:storeId` has potentially up to 500 (because I didn't enforce a store to only fall into one division in my generator, we aren't using division so that shouldn't matter). Since this most queries to this collection include `divId`, `storeId` and `upc` (most queries are for one or more specific products for a specific store) partitioning on `id` as a concatenation of those three properties will give us the best performance for both writes and queries.

As for changing to `divisionId` from `divId`, I'm okay either way. This property is included in a lot of places and is always spelled like that so if we make a change it would need to be made everywhere.

codepossible committed:
Updates to the schema document

codepossible committed:
Updated based on Justine's input

codepossible committed:
Markdown Linting

codepossible commented:
Updated the Solution Design document to encourage exploration, based on discussion with @jcocchi, @davidkpiano and @ianphil.

Split up the message schema document into data store message schema (required) and event schema (suggested).

The happy middle was to provide specific guidance to technology to be user, input and output contacts (define schema for documents in Cosmos DB) but let the participants decide on the exchange of messages between the internal components.

Please review and comment/approve.

jcocchi commented on docs/solution-design.md:
```diff
@@ -6,6 +6,54 @@ While there are many ways to achieve the goal we are looking for, for the purpos
+
+The message schemas define the structure of the messages exchanged between various components.
+
+The [Data Store Message Schema Document](./message-schema-definitions.mdm) provides the structure of the On Hand and Shipment documents and the structure for the record for Master Data Store (MDS). They are essentially the input and output data contracts for the system and are **required** to be used in the defined fashion.
```
fixing broken link

```suggestion
The [Data Store Message Schema Document](./message-schema-definitions.md) provides the structure of the On Hand and Shipment documents and the structure for the record for Master Data Store (MDS). They are essentially the input and output data contracts for the system and are **required** to be used in the defined fashion.
```

codepossible committed:
Update docs/solution-design.md

Co-authored-by: Justine Cocchi <jucocchi@microsoft.com>
